### PR TITLE
docs: update universal guide based on PR 17573

### PR DESCRIPTION
**Pull request by wardbell**
Incorporates all changes from PR #17573; please refer to that PR for earlier history.
Cherry picks to side-step merge commits and conflicts

### CODE REVIEW

File: aio/content/guide/universal.md [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+
+**Angular Universal** generates _static_ application pages on the _server_
+through a process called **server-side rendering (SSR)**.
+
+It can generate and serve those pages in response to requests from browsers.
+It can also pre-generate pages as HTML files that you serve later.
+
+Universal's server-side rendering has several potential benefits:
+
+* [Search Engine Optimization (SEO)](#seo).
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would replace SEO with Web Crawlers then mention SEO, Google, Facebook preview link, twitter preview link etc

File: aio/content/guide/universal.md [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+* [Faster perceived launch](#startup-performance) in many browser scenarios.
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would consider something like "Faster Time to Content" but that's just my personal preference.

File: aio/content/guide/universal.md [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+* [Execution on low power devices or with JavaScript disabled](#no-javascript).
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: Does this mean no Mobile?

File: aio/content/guide/universal.md [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+
+The `renderModuleFactory` renders that view within the `<app>` tag of the template, creating a finished HTML page for the client.
+
+Finally, the server returns the rendered page to the client.
+
+{@a limitations}
+
+### Limitations
+
+Because it doesn't use an actual browser, a Universal `platform-server` app has some limitations:
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would phrase this to say something along the lines of "most Browser APIs we take for granted are not available"

File: aio/content/guide/universal.md [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+The route comes from the client's request to the server.
+Each request results in the appropriate view for the requested route.
+
+The `renderModuleFactory` renders that view within the `<app>` tag of the template, creating a finished HTML page for the client.
+
+Finally, the server returns the rendered page to the client.
+
+{@a limitations}
+
+### Limitations
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: something like "How to write code without Browser APIs" the idea is to not talk in a negative tone because there is a lot a developer needs to learn

File: aio/content/guide/universal.md [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+  index.html                  <i>app web page</i>
+  main.ts                     <i>bootstrapper for client app</i>
+  style.css                   <i>styles for the app</i>
+  systemjs.config.js          <i>SystemJS client configuration</i>
+  systemjs-angular-loader.js  <i>SystemJS add-in</i>
+  tsconfig.json               <i>TypeScript client configuration</i>
+
+  app/  ...                   <i>application code</i>
+
+  uni/                       <i>* folder for universal code</i>
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I did a twitter vote a long time ago. everyone prefers to spell out universal rather than using uni

File: aio/content/guide/universal.md [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+    app-server.module.ts     <i>* server-side application module</i>
+    universal-server.ts      <i>* express web server</i>
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would just call this server since it's only running on the server

File: aio/content/guide/universal.md [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+    universal-engine.ts      <i>* express template engine</i>
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: same (server)

File: aio/content/guide/universal.md [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+rm src/main.js* && rm src/app/*.js*
+</code-example>
+
+<br><hr>
+ 
+{@a server-code}
+
+## Server code
+
+To run an Angular Universal application on the server, you need a server program that accepts client requests and returns rendered pages.
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: server program or platform?

File: aio/content/guide/universal.md [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+{@a webpack-configuration}
+
+### Webpack configuration
+
+Create a `webpack.config.uni.js` file in the project root directory with the following code.
+
+<code-example path="universal/webpack.config.uni.js" title="webpack.config.uni.js">
+</code-example>
+
+**Webpack configuration** is a rich topic beyond the scope of this guide.
```
Comment (gdi2290) [inline aio/content/guide/universal.md]:
```
const excludeNodeModules = require('webpack-node-externals');
  'externals': [excludeNodeModules({
    whitelist: [
      'lit-html'
    ]
  })],
  'target': 'node',
  'node': {
    global: true,
    crypto: true,
    __dirname: true,
    __filename: true,
    process: true,
    Buffer: true
  }

```

File: aio/content/guide/universal.md [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+
+<code-example format="." language="bash">
+rm src/main.js* && rm src/app/*.js*
+</code-example>
+
+</div>
+
+{@a server-app-limitations}
+
+#### Server-rendered app limitations
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would just ban the word limitations

File: aio/content/guide/universal.md [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+
+It's easy to see the limitations of the Universal app when there is no client version of the app.
+
+Navigation via `routerLinks` works correctly.
+You can go from the Dashboard to the Heroes page and back.
+You can click on a hero on the Dashboard page to display its Details page.
+
+But clicks, mouse-moves, and keyboard entries are inert.
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: I would move this section up higher

File: aio/content/guide/universal.md [hunk 13]
```diff
@@ -0,0 +1,823 @@
...
+The server-rendered app still launches quickly but the full client app takes many seconds to load.
+
+{@a conclusion}
+
+## Conclusion
+
+This guide showed you how to take an existing Angular application and make it into a Universal app that does server-side rendering.
+It also showed some of the reasons for doing so:
+
+ - Search Engine Optimization
```
Comment (gdi2290) [inline aio/content/guide/universal.md]: update with stuff mentioned above

**Comment by m98** [comment]
@mary-poppins thanks for your good changes, I noticed a problem and I'm reporting it here (maybe it's just a rare case)
This is still not merged, so I can't pull request on some of the changed files.
I got this error when I was trying to build:

> Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
>  - configuration.output.path: The provided value "src/dist" is not an absolute path!

Here is where the Angular app is located:
```
My-whole-website
├── another-directory
├── angular-website
│   ├── webpack.config.uni.js

```

The problem can be fixed by providing the absolute path of the current directory, we can use` __dirname`, then output's path in the `webpack.config.uni.js` should be:
```
output: {
    path: __dirname+'src/dist',
    filename: 'server.js'
  }
```

In my case just adding `__dirname` helped me to solve the problem.

**Comment by Toxicable** [comment]
Might be better to use the CLI here rather than creating another webpack config
ref: https://github.com/angular/angular-cli/wiki/stories-universal-rendering

**Comment by m98** [comment]
@Toxicable but a webpack is needed anyway

**Comment by Toxicable** [comment]
@m98 not if you use the CLI

**Comment by m98** [comment]
@Toxicable I see you made some changes on [universal-starter](https://github.com/angular/universal-starter), did you try to make a PR?

**Comment by Toxicable** [comment]
@m98 I have, you can check it out here if you like https://github.com/angular/universal-starter/pull/418

File: aio/content/guide/universal.md [hunk 14]
```diff
@@ -0,0 +1,845 @@
...
+Finally, the server returns the rendered page to the client.
+
+{@a limitations}
+
+### Limitations
+
+Because it doesn't use an actual browser, a Universal `platform-server` app has some limitations:
+
+ 1. No browser types such as `window`, `document`, `navigator` or `location`
+ 1. No direct access to the DOM
```
Comment (vikerman) [inline aio/content/guide/universal.md]: Remove this? This is more nuanced with https://github.com/angular/angular/commit/2f2d5f35bd086abeafceabdd5ff9430221b1917f

File: aio/content/guide/universal.md [hunk 15]
```diff
@@ -0,0 +1,845 @@
...
+The core application files are mostly untouched, with a few exceptions described below.
+You'll add more files to support building and serving with Universal.
+
+In this example, Webpack tools compile and bundle the Universal version of the app with the 
+[AOT (Ahead-of-Time) compiler](guide/aot-compiler).
+A node/express web server turns client requests into the HTML pages rendered by Universal.
+
+You will create:
+
+ * a server-side app module, `app.server.ts`
```
Comment (vikerman) [inline aio/content/guide/universal.md]: The filename doesn't match with the one mentioned in the folder structure below

File: aio/content/guide/universal.md [hunk 16]
```diff
@@ -0,0 +1,845 @@
...
+{@a universal-engine}
+
+### Universal template engine
+
+The Universal `renderModuleFactory` function turns a client's requests into server-rendered HTML pages.
+You'll call that function within a _template engine_ that's appropriate for your server stack.
+
+This guide's sample is written for [Node Express](https://expressjs.com/)
+so the engine takes the form of [Express template engine middleware](https://expressjs.com/en/guide/using-template-engines.html).
+
```
Comment (vikerman) [inline aio/content/guide/universal.md]: Mention https://github.com/angular/universal/tree/master/modules/ng-express-engine ?

File: aio/content/guide/universal.md (hunk 6 above)
Comment (vikerman) [reply aio/content/guide/universal.md]: Agreed. universal/ is better for the folder name.

**Review by vikerman — CHANGES_REQUESTED** [review]

File: aio/content/guide/universal.md (hunk 14 above)
Comment (Toxicable) [reply aio/content/guide/universal.md]: I don't think removing it is the best option but changing it to show an example of how you might Inject DOCUMENT would be a good idea

Comment (vikerman) [reply aio/content/guide/universal.md]: Agreed. Let's leave this in for now.

File: aio/content/guide/universal.md (hunk 6 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: Done

File: aio/content/guide/universal.md (hunk 12 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: Done

File: aio/content/guide/universal.md (hunk 16 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: I cannot bring myself to recommend this at this time, at least as far as I understand it, because it does not address critical issues:
1. authentication and unauthorized access to the rendered pages.
1. differentiating static file requests from navigation requests
1. where the client sends requests for data.

The server supplied by the current example doesn't _solve all_ of these problems but it calls them out with placeholders in the express server for you to do something about them.

Do we really want the server to (optionally) compile the module on the server? Shouldn't it be precompiled in the build process as the current guide recommends?

It is not clear from the README how you would use this server. It doesn't look like a "normal" express server as most express developers understand "normal". The repo lacks an example.

I can't see the point of recommending _two_ express servers. If we wanted two, why only showing _one of them in the guide? It's not like the `ngExpressEngine` is self-explanatory.

Finally, it takes a different approach than the guide and we'd have to alter the proposed guide considerably to accommodate it.

This just seems like the wrong direction _for the moment_.

**Comment by wardbell** [comment]
@m98 - regarding your build error, I have no idea how you got it or how to reproduce it. For me the sample builds and runs as documented once the AIO environment as been established.

@Toxicable We made a tactical decision NOT to use the CLI at this time. We're moving toward for the entire suite of documentation and examples.

File: aio/content/guide/universal.md (hunk 16 above)
Comment (Toxicable) [reply aio/content/guide/universal.md]: Thanks for the feedback on this, I'll investigate on how we can improve on those points for future releases

Comment (vikerman) [reply aio/content/guide/universal.md]: Acknowledged

**Review by vikerman — APPROVED** [review]

**Comment by wardbell** [comment]
@gkalpak Looks like you can get us past the remaining approval barrier.

File: aio/content/guide/universal.md (hunk 16 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: @Toxicable To be clear, I am NOT knocking your effort. It would be much better to reference the `ngExpressEngine` (whose direction I very much like) than to rely on the homebrew one presented here.

I'm merely saying that the impact of coping with the issues I raised would postpone this PR beyond what is desired.

Comment (Toxicable) [reply aio/content/guide/universal.md]: @wardbell that's totally understandable, I think that getting it out with something is better than delaying for an extended period of time and meanwhile we can work to resolve those issues and update it then.
If you have time would you mind opening an issue on angular/universal outlining those concerns just so it's a bit easier to track

File: aio/content/examples/.gitignore [hunk 17]
```diff
@@ -62,5 +62,8 @@ dist/
 # styleguide
 !styleguide/src/systemjs.custom.js
 
+# universal
+!universal/**/webpack.config.universal.js
```
Comment (gkalpak) [inline aio/content/examples/.gitignore]: AFAICT, there is only a `universal/webpack.config.universal.js` file, so `/**` is unnecessary (but might be intentionally added).

**Review by gkalpak — APPROVED** [review]
I only skimmed through the guide. I haven't looked at the content and example. The non-content related bits LGTM.

File: aio/content/examples/.gitignore (hunk 17 above)
Comment (wardbell) [reply aio/content/examples/.gitignore]: Fixing next commit

**Comment by wardbell** [comment]
Failing CI is unrelated; both `upgrade-phonecat-1-typescript` and `upgrade-phonecat-2-hybrid` failed e2e

**Comment by wardbell** [comment]
The SECOND COMMIT use webpack to produce the client bundle and moves all served assets to the `dist/` folder.

This better aligns with what you would actually do (you'd serve a production version of the client app, not a chatty, SystemJS dev build). It also simplifies/secures the node express server such that _it only serves out of the `dist/` folder_.

Added notes about the potential role of the CLI in building the client-app as opposed to doing so by hand as this commit does.

**Comment by wardbell** [comment]
@petebd There appears to be a bug in `yarn docs` (and `yarn serve-and-sync` by extension).

It happens for universal PR https://github.com/angular/angular/pull/18707
If I `yarn docs`, it fails with
```
error:   Error processing docs:  Error: Missing example file... relativePath: "universal/webpack.config.client.js". - doc "guide/universal" (content)  - from file "guide/universal.md" - starting at line 1
Example files can be found in: examples, examples
```
This is _emphatically not true_.  The file is there. The example would fail without it.

*SURPRISE!*  If I change the word “client” to “universal” in the path name, `yarn serve-and-sync` succeeds.  Now I go edit `universal.md`, while it’s still watching-and-gening-docs, and change it BACK to “client” … and now it works and shows the correct file in the code pane!

In other words, it _does find the file_ and renders it … but never the first time.

Here’s the link to the doc in the viewer: http://localhost:4200/guide/universal#build-the-client-by-hand

**Comment by gkalpak** [comment]
No surprize there 😛 

I am not sure if it is documented (it should if not), but we have configured dgeni to take `aio/content/examples/.gitignore` into account for collecting example files.

https://github.com/IdeaBlade/angular/blob/c4148b95abd450e5747ec99d8b8208621dbf72ec/aio/tools/transforms/angular-content-package/index.js#L23-L30

All `.js` files are ignored by default (except if explicitly "unignored"):

https://github.com/IdeaBlade/angular/blob/c4148b95abd450e5747ec99d8b8208621dbf72ec/aio/content/examples/.gitignore#L21

Unlike `universal/webpack.config.client.js`, `universal/webpack.config.universal.js` is explicitly "unignored": 

https://github.com/IdeaBlade/angular/blob/c4148b95abd450e5747ec99d8b8208621dbf72ec/aio/content/examples/.gitignore#L67

##
In addition to documenting that, it might be a god idea to add some info about `.gitignore` in the error message itself.

**Comment by petebacondarwin** [comment]
@wardbell I have made a PR to improve the dgeni error message in this case: https://github.com/angular/angular/pull/19265

**Comment by wardbell** [comment]
Thanks for the insight. Have updated the commit accordingly. Hope it passes.

**Comment by wardbell** [comment]
The CI continues to fail for reasons unrelated to this PR; both `upgrade-phonecat-1-typescript` and `upgrade-phonecat-2-hybrid` are failing e2e their tests.

**Review by vikerman — APPROVED** [review]

**Comment by petebacondarwin** [comment]
> The CI continues to fail for reasons unrelated to this PR; both upgrade-phonecat-1-typescript and upgrade-phonecat-2-hybrid are failing e2e their tests.

@wardbell is this PR rebased on the latest master? I seem to remember something about animation errors in builds on master last weekend?

**Comment by wardbell** [comment]
@petebacondarwin I just rebased on current master and force pushed so we'll see.
@juleskremer I made the edits we discussed yesterday. You likely have some more. I'll connect with you on those.

**Comment by petebacondarwin** [comment]
https://travis-ci.org/angular/angular/jobs/277855878#L2317 :-(

**Comment by wardbell** [comment]
I added `pr: cleanup` label because @juleskremer would like to make more prose changes.

File: aio/content/examples/universal/src/universal/server.ts [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+// #docregion data-request
+// TODO: implement data requests securely
+server.get('/api/*', (req, res) => {
+  res.status(404).send('data requests are not supported');
+});
+// #enddocregion data-request
+
+// #docregion navigation-request
+// simplistic regex matches any path without a '.'
+const pathWithNoExt = /^([^.]*)$/;
```
Comment (CaerusKaru) [inline aio/content/examples/universal/src/universal/server.ts]: This is unnecessary, you can just have an `app.use('*.*', express.static('dist'));` for the static files, and then 
```
app.get('*', (req, res) => {
    res.render('index-universal.html', { req });
}
```

File: aio/content/guide/universal.md [hunk 19]
```diff
@@ -0,0 +1,853 @@
...
+{@a seo}
+{@a web-crawlers}
+#### Facilitate web crawlers
+
+Google, Bing, Facebook, Twitter and other social media sites rely on web crawlers to index your application content and make that content searchable on the web. 
+
+These web crawlers may be unable to navigate and index your highly-interactive, Angular application as a human user could do.
+
+Angular Universal can generate a static version of your app that is easy searchable, linkable, and navigable without JavaScript.
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: /s/easy/easily

File: aio/content/guide/universal.md [hunk 20]
```diff
@@ -0,0 +1,853 @@
...
+
+A Universal app can act as a dynamic, content-rich "splash screen" that engages the user.
+It gives the appearance of a near-instant application.
+
+Meanwhile, the browser downloads the client app scripts in background.
+Once loaded, Angular transitions from the static server-rendered page to the dynamically rendered views of the interactive client app.
+
+You must make a few changes to your application code to support both server-side rendering and the transition to the client app.
+
+#### The root `AppModule`
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: You should add a comment on preventing screen "flicker" by adding `initialNavigation: 'enabled'` to a user's root `RouterModule`

File: aio/content/guide/universal.md (hunk 16 above)
Comment (CaerusKaru) [reply aio/content/guide/universal.md]: > authentication and unauthorized access to the rendered pages.

The `renderModuleFactory` doesn't handle this either. It's a function of the client app to handle this browser-side, i.e. with route guards, and for the /api handle to implement authentication/authorization for those guards with a token/cookie

> differentiating static file requests from navigation requests

This is also not handled by the `renderModuleFactory` directly, it's a function of where you specify `server.get('*.*', express.static('dist'))`, i.e. before the rendering process

> where the client sends requests for data.

This is also not handled by the `renderModuleFactory`, it's a function of where you specify `server.get('api/*', ...)`, i.e. before the rendering process

> Do we really want the server to (optionally) compile the module on the server? Shouldn't it be precompiled in the build process as the current guide recommends?

Using the CLI, the module is precompiled, and then used by the ngExpressEngine. You can also pass in the compiled module from your build process. They are equivalent

> It is not clear from the README how you would use this server. It doesn't look like a "normal" express server as most express developers understand "normal". The repo lacks an example.
> I can't see the point of recommending two express servers.

ngExpressEngine is not a server, it is a rendering engine for use with `server.engine('html', ...)`. In this way, it is almost entirely comparable with your method, except it is far cleaner

> It's not like the `ngExpressEngine` is self-explanatory

This is a great reason to clarify its use in this guide

> Finally, it takes a different approach than the guide and we'd have to alter the proposed guide considerably to accommodate it

Again, since it is mostly comparable to `renderModuleFactory`, except clearer, it shouldn't be considerably difficult to add the changes

> The repo lacks an example

The `universal-starter` that @Toxicable mentioned has a fantastic example of incorporating the ngExpressEngine, found [here](https://github.com/angular/universal-starter/blob/master/cli/server.ts)

File: aio/content/guide/universal.md [hunk 21]
```diff
@@ -0,0 +1,853 @@
...
+#### Filter request URLs
+
+The web server must distinguish _app page requests_ from other kinds of requests.
+
+It's not as simple as intercepting a request to the root address `/`.
+The browser could ask for one of the application routes such as `/dashboard`, `/heroes`, or `/detail:12`. 
+In fact, if the app were _only_ rendered by the server, _every_ app link clicked would arrive at the server
+as a navigation URL intended for the router.
+
+Fortunately, application routes have something in common: their URLs lack file extensions.
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: See nit about this in the server file. Should use `express.static` instead of creating unnecessary filters and checks

File: aio/content/guide/universal.md [hunk 22]
```diff
@@ -0,0 +1,853 @@
...
+<div class="alert is-critical">
+
+**Universal HTTP requests have different security requirements**
+
+HTTP requests issued from a browser app are not the same as when issued by the universal app on the server.
+
+When a browser makes an HTTP request, the server can make assumptions about cookies, XSRF headers, etc.
+
+For example, the browser automatically sends auth cookies for the current user.
+Angular Universal cannot forward these credentials to a separate data server.
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: In fact, Angular Universal can and does forward these credentials. The only reason to add additional security plumbing is if the app calls `fetch` (included as a polyfill on the client) for something like GraphQL. This is because Angular Universal doesn't have an implementation, and thus doesn't know to forward the credentials. XHR is totally fine, though.

File: aio/content/guide/universal.md [hunk 23]
```diff
@@ -0,0 +1,853 @@
...
+
+<code-example path="universal/src/universal/server.ts" title="src/universal/server.ts (navigation)" region="navigation-request" linenums="false">
+</code-example>
+
+#### Serve static files safely
+
+A single `server.use()` treats all other URLs as requests for static assets
+such as JavaScript, image, and style files.
+
+To ensure that clients can only download the files that they are _permitted_ to see, you will [put all client-facing asset files in the `/dist` folder](#universal-webpack-configuration)
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: See nit above about using `express.static`

File: aio/content/examples/universal/src/app/hero-search.service.ts [hunk 24]
```diff
@@ -1,20 +1,28 @@
...
 @Injectable()
 export class HeroSearchService {
 
-  constructor(private http: Http) {}
+  private searchUrl = 'api/heroes/?name=';  // URL to web api
+
+  constructor(
+    private http: HttpClient,
+    @Optional() @Inject(APP_BASE_HREF) origin: string) {
+    this.searchUrl = (origin || '') + this.searchUrl;
```
Comment (CaerusKaru) [inline aio/content/examples/universal/src/app/hero-search.service.ts]: Maybe change to ``this.searchUrl = `${origin}${this.searchUrl}` `` to be in line stylistically with other examples, and the very definition of origin in `getOrigin`?

File: aio/content/examples/universal/src/app/hero.service.ts [hunk 25]
```diff
@@ -1,87 +1,64 @@
...
-  // #enddocregion update
-  // #docregion getHeroes
   private heroesUrl = 'api/heroes';  // URL to web api
 
-  constructor(private http: Http) { }
+  // #docregion ctor
+  constructor(
+    private http: HttpClient,
+    @Optional() @Inject(APP_BASE_HREF) origin: string) {
+    this.heroesUrl = (origin || '') + this.heroesUrl;
```
Comment (CaerusKaru) [inline aio/content/examples/universal/src/app/hero.service.ts]: Same nit here as above

File: aio/content/examples/universal/src/app/hero.service.ts [hunk 26]
```diff
@@ -1,87 +1,64 @@
...
-import { Headers, Http } from '@angular/http';
+import { Injectable, Inject, Optional } from '@angular/core';
+import { APP_BASE_HREF } from '@angular/common';
+import { HttpHeaders, HttpClient } from '@angular/common/http';
 
-// #docregion rxjs
-import 'rxjs/add/operator/toPromise';
-// #enddocregion rxjs
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/catch';
```
Comment (CaerusKaru) [inline aio/content/examples/universal/src/app/hero.service.ts]: This is an RxJS anti-pattern. Please use `.call` on each operator, or wait until Angular supports RxJS 5.5.0 and its lettable operators

File: aio/content/guide/universal.md [hunk 27]
```diff
@@ -0,0 +1,853 @@
...
+ * `@types/express` - TypeScript type definitions for express.
+
+Install them with the following commands:
+
+<code-example format="." language="bash">
+npm install @angular/compiler-cli @angular/platform-server express --save
+npm install webpack @ngtools/webpack copy-webpack-plugin raw-loader @types/express --save-dev
+</code-example>
+
+{@a transition}
```
Comment (CaerusKaru) [inline aio/content/guide/universal.md]: Add a section on capturing user processes using `@angular/preboot`? This should probably be a side note, because getting preboot to work with the current example would require significant refactoring

**Review by CaerusKaru — CHANGES_REQUESTED** [review]
I think it's a big mistake to not include out-of-the-box support for Angular CLI and ngExpressEngine. Not including these widely-used tools may lead to user confusion and bad practice out of the gate. [Many](https://davidea.st/articles/the-beginners-guide-to-angular-universal) of the [guides](https://medium.com/@cyrilletuzi/angular-server-side-rendering-in-node-with-express-universal-engine-dce21933ddce) use these tools and they make for a clearer process. I don't think it would require a lot of refactoring to get this guide in line with those tools, and I would volunteer to do the work myself if it's too much, @wardbell.

I understand the desire to get this guide out ASAP, but I think it's best to hold off, since there are other guides out there in the interim, if it means getting the best guide possible.

All in all though, excellent work putting this together @wardbell, and I don't mean to be harsh.

File: aio/content/examples/universal/src/app/hero.service.ts (hunk 26 above)
Comment (Foxandxss) [reply aio/content/examples/universal/src/app/hero.service.ts]: Explain? I am interested.

Comment (CaerusKaru) [reply aio/content/examples/universal/src/app/hero.service.ts]: See https://github.com/angular/material2/issues/2622

**Comment by wardbell** [comment]
@CaerusKaru We discussed timing internally and decided to forge ahead and revisit as quickly as alternatives allow. It's a judgment call and not mine to make.

I'll look at your other comments when I have time.

File: aio/content/guide/universal.md (hunk 27 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: I was asked to remove the reference to preboot that I had in an early version of this guide.

File: aio/content/examples/universal/src/app/hero.service.ts (hunk 26 above)
Comment (wardbell) [reply aio/content/examples/universal/src/app/hero.service.ts]: @CaerusKaru This is an anti-pattern in a _library_ because you risk screwing the consumer of your library, someone you do not and cannot know.

Patching the prototype is OK in your own _application_. You're not screwing anyone but yourself.  The `call` syntax is bloody awful. It completely destroys the dev experience. Save it for libraries.

The library/app difference is a critically important distinction!

I do hope we all move to lettable when it releases. But it is still in beta as I write this.

File: aio/content/guide/universal.md (hunk 22 above)
Comment (wardbell) [reply aio/content/guide/universal.md]: I'm parroting what I believe I was told by @alxhub

File: aio/content/examples/universal/src/app/hero.service.ts (hunk 26 above)
Comment (Toxicable) [reply aio/content/examples/universal/src/app/hero.service.ts]: Sadly there's a side effect of even using the `lettable` method of rxjs right now, it's that if you use that method along side the prototype patching method then you'll end up with duplicated operators in your bundle.
From memory Webpack will solve this in v4 but for now it kinda of stops any libs from using it.
Sorry I don't have a think for this, just recalling it from some issues I've read

Comment (CaerusKaru) [reply aio/content/examples/universal/src/app/hero.service.ts]: I agree with the reasoning for libraries, but it's also a benefit for tree-shaking. By patching when you need it, you avoid patching everywhere.

I also agree that the `call` syntax is terrible, I wanted to leave this note here primarily for when lettable operators are usable in Angular core.

File: aio/content/guide/universal.md (hunk 22 above)
Comment (CaerusKaru) [reply aio/content/guide/universal.md]: It may be that I'm using ngExpressEngine, which injects the `req` and `res` objects from Express by default. Regardless, it's definitely possible to add this functionality in only a couple of lines of code.

Here's the snippet from ngExpressEngine:
https://github.com/angular/universal/blob/2a1246326d5db10a989f143205c2d2ed27f8e362/modules/express-engine/src/main.ts#L129-L143

**Comment by chuckjaz** [comment]
@wardbell This PR does not apply cleanly on 4.4.x. Can you create a 4.4.x PR with these changes?

**Comment by wardbell** [comment]
@chuckjaz I believe @vikerman is dealing with the 4.4 problems.
